src/osd/PGLog.h

```diff
@@ -1057,27 +1057,7 @@ struct PGLog : DoutPrefixProvider {
       /// Case 1)
       ldpp_dout(dpp, 10) << __func__ << ": more recent entry found: "
 			 << *objiter->second << ", already merged" << dendl;
-
-      ceph_assert(objiter->second->version > last_divergent_update);
-
-      // ensure missing has been updated appropriately
-      if (objiter->second->is_update() ||
-	  (missing.may_include_deletes && objiter->second->is_delete())) {
-	ceph_assert(missing.is_missing(hoid) &&
-	       missing.get_items().at(hoid).need == objiter->second->version);
-      } else {
-	ceph_assert(!missing.is_missing(hoid));
-      }
-      missing.revise_have(hoid, eversion_t());
-      missing.mark_fully_dirty(hoid);
-      if (rollbacker) {
-	if (!object_not_in_store) {
-	  rollbacker->remove(hoid);
-	}
-	for (auto &&i: entries) {
-	  rollbacker->trim(i);
-	}
-      }
+      ceph_abort_msg("unexpected error, actually we shouldn't have gotten here!");
```

Comment:
You also need to update the explanation at the top as this is no longer a valid case...  The goal here is for both of us and every future reader of this code to have a very clear idea of what is possible and why.

Reply:
Though, you'll want to have a look at my other comments first.